### PR TITLE
Added a clone descriptor method.

### PR DESCRIPTION
**Pull request by cjqian**
This makes a deep copy of a given tag helper descriptor.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Razor.Runtime/Runtime/TagHelpers/TagHelperDescriptorResolver.cs [hunk 1]
```diff
@@ -143,6 +143,35 @@ public IEnumerable<TagHelperDescriptor> Resolve(TagHelperDescriptorResolutionCon
             return descriptors;
         }
 
+        /// <summary>
+        /// This method returns a deep copy of a given <see cref="TagHelperDescriptor"/>.
+        /// </summary>
+        /// <param name="descriptor">The <see cref="TagHelperDescriptor"/> to clone.</param>
+        /// <returns></returns>
+        public static TagHelperDescriptor CloneDescriptor(TagHelperDescriptor descriptor)
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Razor.Runtime/Runtime/TagHelpers/TagHelperDescriptorResolver.cs]: Would this be better suited as an instance method on `TagHelperDescriptor` i.e. `public TagHelperDescriptor Clone()`? Seems odd to have this be a public method on the resolver.

File: src/Microsoft.AspNetCore.Razor.Runtime/Runtime/TagHelpers/TagHelperDescriptorResolver.cs [hunk 2]
```diff
@@ -143,6 +143,35 @@ public IEnumerable<TagHelperDescriptor> Resolve(TagHelperDescriptorResolutionCon
...
+        {
+            var cloneDescriptor = new TagHelperDescriptor
+            {
+                Prefix = descriptor.Prefix,
+                TagName = descriptor.TagName,
+                TypeName = descriptor.TypeName,
+                AssemblyName = descriptor.AssemblyName,
+                Attributes = descriptor.Attributes,
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Razor.Runtime/Runtime/TagHelpers/TagHelperDescriptorResolver.cs]: Would the attributes need to be deep copied too (the same way the `PropertyBag` is)?

**Comment by cjqian** [comment]
:up: :date:

File: src/Microsoft.AspNetCore.Razor/Compilation/TagHelpers/TagHelperDescriptor.cs [hunk 3]
```diff
@@ -24,6 +24,37 @@ public class TagHelperDescriptor
...
         /// <summary>
+        /// Creates a new <see cref="TagHelperDescriptor"/>. 
+        /// </summary>
+        public TagHelperDescriptor()
+        {
+
+        }
+
+        /// <summary>
+        /// Creates a copy of the given <see cref="TagHelperDescriptor"/>. Note: this is not a deep copy.
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor/Compilation/TagHelpers/TagHelperDescriptor.cs]: `Creates a shallow copy of the given <see cref="TagHelperDescriptor"/>.`

Nuke the `Note: this is not a deep copy.`

File: src/Microsoft.AspNetCore.Razor/Compilation/TagHelpers/TagHelperDescriptor.cs [hunk 4]
```diff
@@ -24,6 +24,37 @@ public class TagHelperDescriptor
...
+            TypeName = descriptor.TypeName;
+            AssemblyName = descriptor.AssemblyName;
+            Attributes = descriptor.Attributes;
+            RequiredAttributes = descriptor.RequiredAttributes;
+            AllowedChildren = descriptor.AllowedChildren;
+            RequiredParent = descriptor.RequiredParent;
+            TagStructure = descriptor.TagStructure;
+            DesignTimeDescriptor = descriptor.DesignTimeDescriptor;
+
+            foreach (var property in descriptor.PropertyBag)
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor/Compilation/TagHelpers/TagHelperDescriptor.cs]: You can actually just set the `_propertyBag` field.

**Comment by cjqian** [comment]
:up: :date:

File: src/Microsoft.AspNetCore.Razor/Compilation/TagHelpers/TagHelperDescriptor.cs [hunk 5]
```diff
@@ -24,6 +24,37 @@ public class TagHelperDescriptor
             Enumerable.Empty<TagHelperRequiredAttributeDescriptor>();
 
         /// <summary>
+        /// Creates a new <see cref="TagHelperDescriptor"/>. 
+        /// </summary>
+        public TagHelperDescriptor()
+        {
+
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor/Compilation/TagHelpers/TagHelperDescriptor.cs]: nit: Nuke extra line.

File: test/Microsoft.AspNetCore.Razor.Test/Compilation/TagHelpers/TagHelperDescriptorTest.cs [hunk 6]
```diff
@@ -13,6 +13,47 @@ namespace Microsoft.AspNetCore.Razor.Compilation.TagHelpers
...
+                DesignTimeDescriptor = new TagHelperDesignTimeDescriptor()
+            };
+
+            descriptor.PropertyBag.Add("foo", "bar");
+
+            // Act
+            var copyDescriptor = new TagHelperDescriptor(descriptor);
+
+            // Assert
+            Assert.Equal(descriptor, copyDescriptor, TagHelperDescriptorComparer.Default);
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNetCore.Razor.Test/Compilation/TagHelpers/TagHelperDescriptorTest.cs]: Add asserts down here ensuring Attributes and RequiredAttributes are the same: `Assert.Same`

**Comment by NTaylorMullen** [comment]
:watch:

**Comment by cjqian** [comment]
:up: :date:

File: test/Microsoft.AspNetCore.Razor.Test/Compilation/TagHelpers/TagHelperDescriptorTest.cs (hunk 6 above)
Comment (NTaylorMullen) [inline test/Microsoft.AspNetCore.Razor.Test/Compilation/TagHelpers/TagHelperDescriptorTest.cs]: `CaseSensntiveTagHelperDescriptorComparer`

**Comment by NTaylorMullen** [comment]
Once [this](https://github.com/aspnet/Razor/pull/827/files#r77082392) is done :shipit:

**Pull request merged**